### PR TITLE
[feature-threads] test read after write MOD-4813

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR adds Read after write tests for background threads
It allows to control the execution flow via PTRACE and named breakpoint in the code.
It introduces `SearchEnv` that allows async command execution for testing.
Also, it introduces dedicated GH action job.

How to run:
```
make DEBUG=1 BUILD_TRACE=1
make pytest TEST=test_read_after_writes.py DEBUG=1
```
Note there is the need to build with `DEBUG` flag on.

Discussion:
1. Is this code "good enough" for the time being, with respect to PTRACE and the usage of named breakpoints?
2. Async cmd - move to RLTest?

### CODE REVIEW

File: deps/thpool/thpool.c [hunk 1]
```diff
@@ -103,55 +104,60 @@ static void bsem_wait(struct bsem* bsem_p);
...
-    return NULL;
-  }
+    /* Make threads in pool */
+    thpool_p->threads = (struct thread**)rm_malloc(num_threads * sizeof(struct thread*));
+    if (thpool_p->threads == NULL) {
+      err("thpool_init(): Could not allocate memory for threads\n");
+      jobqueue_destroy(&thpool_p->jobqueue);
+      rm_free(thpool_p);
+      return NULL;
+    }
```
Comment (GuyAv46) [inline deps/thpool/thpool.c:137]: we should also free the name here.
Add the name free to the destructor as well

File: src/util/trace_bp.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+#ifdef BUILD_TRACE
+#define NAMED_TRACED_BP(label) \
+    do { \
+    printf("%s\r\n", label); \
+    register char*  arg2 asm("rsi") = (char*)label; \
+    asm("int3"); \
+    } while(0);
+#else 
+#define NAMED_TRACED_BP(label) ;
```
Comment (GuyAv46) [inline src/util/trace_bp.h:11]: this can be just blank no?

File: src/util/trace_bp.h [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#ifdef BUILD_TRACE
+#define NAMED_TRACED_BP(label) \
+    do { \
+    printf("%s\r\n", label); \
+    register char*  arg2 asm("rsi") = (char*)label; \
+    asm("int3"); \
+    } while(0);
```
Comment (GuyAv46) [inline src/util/trace_bp.h:9]: here we don't have to include the `;` to ensure the user will add it himself

File: src/util/trace_bp.h [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#ifdef BUILD_TRACE
+#define NAMED_TRACED_BP(label) \
+    do { \
+    printf("%s\r\n", label); \
+    register char*  arg2 asm("rsi") = (char*)label; \
```
Comment (GuyAv46) [inline src/util/trace_bp.h:7]: consider adding a comment or a link to some explanation for this code

File: tests/pytests/common.py [hunk 5]
```diff
@@ -370,3 +379,92 @@ def expect_when(self, cond_val, func: Callable[[Query], Any]):
         if cond_val == self.cond_val:
             func(self.env.expect(*self.query))
         return self
+
+def module_path():
+    # path/to/RediSearch/tests/pytests
+    root_path =  dirname(dirname(dirname(os.path.realpath(__file__))))
```
Comment (GuyAv46) [inline tests/pytests/common.py:385]: is this like `ls ../../..`?

File: tests/pytests/test_raw.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
+from RLTest import Env
+from common import *
+from includes import *
+
+@SearchTest(readers = 1)
+def test_sanity(searchEnv):
+    if not DEBUG:
+        searchEnv.skip()
+    searchEnv.expect('ft.create', 'idx', 'ON', 'HASH', 'ASYNC', 'schema', 'name', 'text').ok()
```
Comment (GuyAv46) [inline tests/pytests/test_raw.py:9]: why `ASYNC`?

**Review by GuyAv46 — COMMENTED** [review]
Nice! Few comments
Seems like the bot has a lot to say, did you go over it?

**Review by GuyAv46 — APPROVED** [review]

File: tests/pytests/common.py [hunk 7]
```diff
@@ -370,3 +379,92 @@ def expect_when(self, cond_val, func: Callable[[Query], Any]):
...
+    search_file_path = join(root_path, 'bin', f'{OS}-{ARCH}-{BUILD_MODE}', 'search', 'redisearch.so' )
+    json_file_path = join(root_path, 'bin', f'{OS}-{OSNICK}-{ARCH}', 'RedisJSON', 'rejson.so' )
```
Comment (rafie) [inline tests/pytests/common.py:387]: `runtests.sh` has `MODULE` and `REJSON_PATH` env vars for module paths. Those can we exported to be available for tests.

Comment (GuyAv46) [reply tests/pytests/common.py:387]: what about `ROOT` directory?

**Pull request closed without merging**